runtime/stdlib/cadence_v0.42_to_v1_contract_upgrade_validator.go

```diff
@@ -338,11 +338,31 @@ typeSwitch:
 			return nil
 		}
 	case *ast.IntersectionType:
-		// intersection types cannot be upgraded unless they have a legacy restricted type,
-		// in which case they must be upgraded according to the migration rules: i.e. R{I} -> R
+		// If the intersection type have doesn't a legacy restricted type,
+		// the interface set must be compatible.
 		if oldType.LegacyRestrictedType == nil {
-			break
+			foundIntersectionType, ok := newType.(*ast.IntersectionType)
+			if !ok {
+				return newTypeMismatchError(oldType, newType)
+			}
+
+			if len(oldType.Types) != len(foundIntersectionType.Types) {
+				return newTypeMismatchError(oldType, newType)
+			}
```

Comment:
With interface inheritance, I guess there could be cases where the new set is smaller, but has semantically equal "contents"?  Just a side-note, we do not need to address this here

Reply:
yeah, for now this only allows changing the known "moved" interfaces. I'll see if we can support reducing too, as long as they are semantically equivalent, as a follow-up PR. We would need to make sure it is consistent to how the migration works.

Reply:
Yeah, let's see how this works and we can investigate if needed 👍